tests/gas_subsidies/deducts_funds_test.go

```diff
@@ -0,0 +1,372 @@
+				require.NoError(t, err)
+
+				receipts, err := net.GetReceipts([]common.Hash{sponsoredTx.Hash(), normalTx.Hash()})
+				require.NoError(t, err)
+				require.Equal(t, types.ReceiptStatusSuccessful, receipts[0].Status)
+				require.Equal(t, types.ReceiptStatusSuccessful, receipts[1].Status)
+			},
+		},
+		"sponsored transaction calling a contract which aborts": {
+			runTransactions: func(t *testing.T, net *tests.IntegrationTestNet, sender *tests.Account) {
+
+				opts, err := net.GetTransactOptions(sender)
+				require.NoError(t, err)
+
+				opts.GasLimit = 300_000
+				opts.GasPrice = big.NewInt(0)
+				tx, err := revertContract.DoCrash(opts)
+				require.NoError(t, err)
+				receipt, err := net.GetReceipt(tx.Hash())
+				require.NoError(t, err)
+				require.Equal(t, types.ReceiptStatusFailed, receipt.Status)
+			},
+		},
+	}
+
+	for name, test := range cases {
+		t.Run(name, func(t *testing.T) {
+
+			donation := big.NewInt(1e18)
+
+			sponsoredSender := tests.MakeAccountWithBalance(t, net, big.NewInt(1))
+			sponsorshipRegistry := Fund(t, net, sponsoredSender.Address(), donation)
+
+			subsidiesRegistryBalanceBefore, err := client.BalanceAt(t.Context(), registry.GetAddress(), nil)
+			require.NoError(t, err)
+
+			blockBefore, err := client.BlockByNumber(t.Context(), nil)
+			require.NoError(t, err)
+
+			// Here the scenario is ran.
+			test.runTransactions(t, net, sponsoredSender)
+
+			blockAfter, err := client.BlockByNumber(t.Context(), nil)
+			require.NoError(t, err)
+
+			config, err := sponsorshipRegistry.GetGasConfig(nil)
+			require.NoError(t, err)
+
+			// Scenarios may produce multiple blocks, so we need to
+			// iterate through all of them to find all sponsored transactions.
+			var fundsDelta uint64
+
+			// For every block created during test scenario
+			for blockNumber := blockBefore.NumberU64() + 1; blockNumber <= blockAfter.NumberU64(); blockNumber++ {
+
+				block, err := client.BlockByNumber(t.Context(), big.NewInt(int64(blockNumber)))
+				require.NoError(t, err)
+
+				var totalGasUsed uint64
+
+				for i, tx := range block.Transactions() {
+					receipt, err := net.GetReceipt(tx.Hash())
+					require.NoError(t, err)
+
+					totalGasUsed += receipt.GasUsed
+					require.EqualValues(t, totalGasUsed, receipt.CumulativeGasUsed)
+
+					if subsidies.IsSponsorshipRequest(tx) {
+						fundsUsed := (receipt.GasUsed +
+							config.OverheadCharge.Uint64()) * block.BaseFee().Uint64()
+						require.Greater(t, fundsUsed, uint64(0),
+							"sponsored tx must have a non-zero funds cost",
+						)
+						fundsDelta += fundsUsed
+
+						require.Less(t, i, len(block.Transactions())-1,
+							"sponsored tx should not be the last tx in the block")
+						internalTx := block.Transactions()[i+1]
+
+						deduceFundsReceipt, err := net.GetReceipt(internalTx.Hash())
+						require.NoError(t, err)
+						require.Equal(t, types.ReceiptStatusSuccessful, deduceFundsReceipt.Status)
+
```

Comment:
Maybe call `validateSponsoredTxInBlock` here, just to get the full features covered by this utility covered here as well?